src/benchmarks/micro/coreclr/perflab/CastingPerf.cs

```diff
@@ -373,5 +375,93 @@ public bool CheckArrayIsInterfaceNo()
                 res = myClass2Arr is IMyInterface1[];
             return res;
         }
+
+        [GlobalSetup(Target = nameof(CheckArrayIsNonvariantGenericInterface))]
+        public void SetupCheckArrayIsNonvariantGenericInterface() => myObj = new MyClass2[5];
+
+        [Benchmark]
+        public bool CheckArrayIsNonvariantGenericInterface()
+        {
+            bool res = false;
+            for (int i = 0; i < InnerIterationCount; i++)
+                res = myObj is ICollection<MyClass2>;
```

Comment:
This does not look like a very robust benchmark pattern. What does prevent the JIT from CSEing this whole thing in front of the loop?

Reply:
I guess nothing. We will see an improvement on the benchmark then. 
I asked myself this question the same question, but this is a pattern used in basically all other casting benchmarks.

I suppose if we really get smart enough to CSE this, we would need to update most of them.

Reply:
The other concern is that JIT may one day move `is` outside of the loop (since `is` is idempotent).

@adamsitnik - any concerns about this?

Reply:
I guess, calling some noninlineable method in the loop will prevent such optimizations. - As long as JIT does not look into that method and believes there could be sideeffects.

That will add overhead though and will certainly have impact on existing benchmarks if we do this to them.

Reply:
I have added calls to `FakeSideEffects()` - a method that does nothing. That should prevent potential CSE.
It looks like the calls do not add a lot of cost compared to casts (at least on my machine), but they do seem add some noise.

I only modified the new scenarios since there is no history of results yet.

Reply:
@jkotas - does this address your concerns?

Reply:
@jkotas - doe this address your concerns

Reply:
It does address the problem, but it does not follow the best practices for writing micro-benchmarks using Benchmark.NET: https://github.com/dotnet/performance/blob/master/docs/microbenchmark-design-guidelines.md

- Microbenchmarks should not have loops. Benchmark.NET does it run in a loop for you.
- The optimization of side effects should be avoided by returning the result from the benchmark method

Reply:
From existing tests in this file I assumed that casts are exception from the "no loops" rule. It is plausible because casts are cheap (compared to a delegate invoke from the outer loop).   

I will try removing loops and see if results are similar and not too noisy.

Reply:
> casts are exception 

I do not see why casts should be an exception. AFAIK, the only exception are ref-types where Benchmark.NET cannot do caching around the callsite efficiently today.

> I only modified the new scenarios since there is no history of results yet.

I do not think we should be keeping poorly written benchmarks around. I think we should fix it and rename the benchmark to avoid false alarm from automated trend analysis (e.g. like here: https://github.com/dotnet/performance/pull/892/)

Reply:
When benchmarks were ported to BDN from xunit.performance, we made the choice (and have since agreed to keep) the existing benchmarks. New benchmarks should not follow this style, however. #189 has some discussion of a set of these.

Please modify the new cases to use the best practices Jan linked to. If doing so results in benchmarks that are unstable in some way, we'll dig into that.

Reply:
I have converted the new benchmarks to the "right style" and rerun on mu machine.

I see some "smoothing" of diffs. - I.E. where before one implementation was taking 1.8x longer than another it now takes 1.6x longer. This mostly impacts cases where differences are small and in a way is a good thing.

I did not see additional noise.